### ISSUE TITLE
[DX] A better upgrade process

### ISSUE THREAD

**Issue by wouterj** [symfony/symfony-standard#730]
## The Current Situation

When a new minor version is released, updating Symfony is simple:

```
$ composer update
```

This will update the `symfony/symfony` package and all other packages. Great!
## The Problem

There is one problem: The edition you're using is unchanged. Changes in the SE are not taking into account while upgrading.

In previous versions, this wasn't a big problem: Only minor things were introduced in the SE. Now 2.6 introduces the VarDumper component _and DebugBundle_ (mostly the last one is causing problems), this will change.

@flip111 and me have been debugging a problem for several hours saturday night: `{{ dump() }}` didn't show anything. After a long debugging process, we found out that we simply forgot to register the DebugBundle into the app kernel.

Today, we had another problem with the debug bundle: We couldn't use `dump()` in our PHP code. It turned out that this too was inlined into the Standard Edition.

All in all, I came to the conclusion that we have to change the update process. Mostly because there are lots of questions about this on IRC after a new version is released and with this new edition, I think we get even more questions.
## The Solution

A while ago, I proposed to also provide a PATCH file for the Standard Edition. People could apply this patch file to their project and the SE will be up to date.

Now we have a new Symfony Installer, I think it's good to put an `upgrade` command in there which does this for you: It checks your files, it checks the changes in the SE and it updates or displays the changes (in case there are conflicts)

**Comment by layanto** [comment]
One way to see what has changed between versions is to use Github compare.

https://github.com/symfony/symfony-standard/compare/2.6...2.7

**Comment by MaxVandervelde** [comment]
@layanto For the lazy, github will even format those as a patch for you if you append `.patch` to the URL: https://github.com/symfony/symfony-standard/compare/2.6...2.7.patch

But I'd go so far as saying that *most* of the time these don't apply cleanly. So this isn't going to be easy. Even though it is a good idea, it would certainly be easier to walk the user through the upgrade steps like @flip111 suggested, instead. Maybe that's a better first step?

**Comment by weaverryan** [comment]
@WouterJ Now that you've added the really nice upgrade docs, could those be enough? Especially if we did a little bit of a better job with the `UPGRADE-*.md` files inside the SE (and mentioned those from there)?

**Comment by ReservedDeveloper** [comment]
Resurrecting this thread: Has anyone developed a better procedure than either: a) cloning the standard and copying in changes, or b) adding a `git remote` and merging in the changes?

**Comment by Pierstoval** [comment]
Each app is so specific that the only thing that could be "standardized" seem to be a list of recommandations. 
I don't know if there is a solution for this :confused:

**Comment by tobymackenzie** [comment]
**TL;DR**: You could: 
1. get the diff between the starting and ending versions, like '2.6...2.7', probably only paying attention to certain files and possibly diffing 'yml' and 'json' files as associative arrays instead of lines
2. generate a diff between that diff and the current project
3. interactively ask if each change should be applied

This may be complicated.

**The long**:
I was working on a bundle to attempt to manage this sort of thing, but kind of gave up.  My initial approach was to have a matching version for each version of the SE that had all composer requirements, a method to create an array of all the bundles, the configuration, and anything else to store as many of the things that could change in SE, so that my projects would just build to this interface and then be able to change versions with a single line composer change.  

Besides for composer root versus non-root issues and others, I realized that not all projects need all of the dependencies and starting configuration of SE, so I sought I different route.  I thought of various possibilities, and began leaning toward a command / script to automatically make changes to the given project.  I began thinking of a two-stage diffing setup where the existing and target SE versions are diffed (certain files only) and then that diff is compared to the current project to come up with what needs changed.  Stuff like 'composer.json' and '.yml' files would be compared specially by diffing array representations instead of line numbers to get at actual configuration values so they wouldn't be affected by significant changes to the files.  Since a given project may remove some SE stuff and not want it automatically added back in, I decided it made sense to have it show each change and confirm whether it should or shouldn't be applied.  It was a bit complicated for me for the perceived value, so I didn't get very far with it.  

I did get somewhere with a tool to generate diffs between associative arrays, if you decide to go down that route.

**Comment by iltar** [comment]
I don't use the SE so I don't have this specific issue, but I do base my `app.php` and `app_dev.php` based on the SE, I manually have to check if those files are changed. I think this falls under the same category of "updates".

**Comment by teohhanhui** [comment]
Even with 2.7 -> 2.8 I got stung by the lack of visibility of the upgrade instructions. Perhaps the Symfony release blog post should link to the upgrade notes for SE as well (not just Symfony proper).

**Comment by sagikazarmark** [comment]
I fought the exact same issues recently. My solution was creating my own, common [standard](https://github.com/sagikazarmark/symfony-standard) repo and using it as a boilerplate. I also preserved the history in projects, so that I can easily merge changes.

**Comment by fabpot** [comment]
Closing as this repo is now closed for feature addition. Flex is the future of Symfony.
